File: src - java/Esercizi/Ribalta/Ribalta_ricorsivo.java
	//Versione ricorsiva del metodo ribalta()

	//Metodo privato
	private Record _ribalta(Record p) {
		// Se la lista contiene 0 o 1 elementi, non faccio nulla.
		if (p == null || p.succ == null)
			return p;

		// memorizzo in vecchiaTesta la vecchia testa
		Record vecchiaTesta = p;
		// memorizzo in vecchioSecondo il vecchio sec. elemento
		Record vecchioSecondo = p.succ;
		// ribalto la sottolista con in testa il vecchio sec...
		// ...e memorizzo in nuovaTesta la nuova testa.
		Record nuovaTesta = _ribaltaRicorsivo(vecchioSecondo);

		// la vecchia testa diviene l'ultimo elemento (e quindi
		// punta a null)
		vecchiaTesta.succ = null;
		// il vecchio secondo elemento punta alla vecchia testa
		vecchioSecondo.succ = vecchiaTesta;

		return nuovaTesta; // restituisco la nuova testa
	}

	//Metodo pubblico
	public void ribalta() {
		first = _ribaltaRicorsivo(first);
	}

File: src - c++/Esercizi/Ribalta/Ribalta_ricorsivo.cpp
//Versione ricorsiva del metodo Ribalta()

//Metodo privato
PRec Lista::_ribalta(PRec p) {
  if ((!p) || (!p->succ)) //se la lista è formata da 0 o 1 elementi
                          //non faccio niente
    return p;
  else {
    //memorizzo in vecchia_testa la vecchia testa
    PRec vecchia_testa = p;
    //memorizzo in vecchio_secondo il vecchio secondo elemento
    PRec vecchio_secondo = p->succ;
    //ribalto la sottolista con testa in vecchio_secondo...
    //...e memorizzo in nuova_testa la nuova testa.
    PRec nuova_testa = _ribalta(vecchio_secondo);
                                  
    //la vecchia testa diviene l'ultimo elemento (e quindi punta a 0)                                  
    vecchia_testa->succ = 0;
    //il vecchio secondo elemento punta alla vecchia testa
    vecchio_secondo->succ = vecchia_testa;
    
    return nuova_testa; //restituisco la nuova testa
  }    
}    

//Metodo pubblico
void Lista::Ribalta() {
  first = _ribalta(first);
}
